crates/sqlbuiltins/src/functions/conversion.rs

```diff
@@ -0,0 +1,212 @@
+use std::fmt;
+
+use datafusion::scalar::ScalarValue;
+
+use super::*;
+/// Value from a function parameter.
+#[derive(Debug, Clone)]
+pub enum FuncParamValue {
+    /// Normalized value from an ident.
+    Ident(String),
+    /// Scalar value.
+    Scalar(ScalarValue),
+    Array(Vec<FuncParamValue>),
+}
+
+impl fmt::Display for FuncParamValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Ident(s) => write!(f, "{s}"),
+            Self::Scalar(s) => write!(f, "{s}"),
+            Self::Array(vals) => write!(f, "{vals:?}"),
+        }
+    }
+}
+
+impl FuncParamValue {
+    /// Print multiple function parameter values.
+    pub fn multiple_to_string<T: AsRef<[Self]>>(vals: T) -> String {
+        let mut s = String::new();
+        write!(&mut s, "(").unwrap();
+        let mut sep = "";
+        for val in vals.as_ref() {
+            write!(&mut s, "{sep}{val}").unwrap();
+            sep = ", ";
+        }
+        write!(&mut s, ")").unwrap();
+        s
+    }
+
+    /// Wrapper over `FromFuncParamValue::from_param`.
+    pub(super) fn param_into<T>(self) -> Result<T>
+    where
+        T: FromFuncParamValue,
+    {
+        T::from_param(self)
+    }
+}
+
+pub(super) trait FromFuncParamValue: Sized {
+    /// Get the value from parameter.
+    fn from_param(value: FuncParamValue) -> Result<Self>;
+
+    /// Check if the value is valid (able to convert).
+    fn is_param_valid(value: &FuncParamValue) -> bool;
+}
+
+impl FromFuncParamValue for String {
+    fn from_param(value: FuncParamValue) -> Result<Self> {
+        match value {
+            FuncParamValue::Scalar(ScalarValue::Utf8(Some(s))) => Ok(s),
+            other => Err(BuiltinError::UnexpectedArg {
+                param: other,
+                expected: DataType::Utf8,
+            }),
+        }
+    }
+
+    fn is_param_valid(value: &FuncParamValue) -> bool {
+        matches!(value, FuncParamValue::Scalar(ScalarValue::Utf8(Some(_))))
+    }
+}
+
+impl FromFuncParamValue for Vec<String> {
+    fn from_param(value: FuncParamValue) -> Result<Self> {
+        match value {
+            FuncParamValue::Array(arr) => {
+                let mut res = Vec::with_capacity(arr.len());
+                for val in arr {
+                    res.push(String::from_param(val)?);
+                }
+                Ok(res)
+            }
+            other => Err(BuiltinError::UnexpectedArg {
+                param: other,
+                expected: DataType::List(Arc::new(Field::new("list", DataType::Utf8, false))),
+            }),
+        }
+    }
+
+    fn is_param_valid(value: &FuncParamValue) -> bool {
+        matches!(value, FuncParamValue::Scalar(ScalarValue::Utf8(Some(_))))
+    }
+}
+
```

Comment:
I'm not sure if it's allowed but can we implement this for `Vec<T> where T: FromFuncParamValue` instead?

Reply:
i think we'd need to add some additional logic to map it to a dtype for `is_param_valid` to work.

Reply:
Yeah. Just noticing that the current logic would fail as well. We do need additional logic anyways.